### PR TITLE
[Impeller] use SSBOs for gradients where supported (metal/vulkan)

### PR DESCRIPTION
**Pull request by jonahwilliams**
Use SSBO to pass the gradient color buffer to the shader, which dramatically improves performance for small gradients. When creating this color buffer, reusing the existing color buffer if the stops are even.

Fixes https://github.com/flutter/flutter/issues/115208 (for iOS).
Fixes https://github.com/flutter/flutter/issues/115523

On Android, any usage of arrays in the fragment shaders leads to a link error with no other messages. Assuming that we're using the gles fragment shaders, I'm not really sure what the issue is exactly - but I assume this has something to do with targeting gles 2.0.

### CODE REVIEW

File: impeller/entity/contents/content_context.cc [hunk 1]
```diff
@@ -157,6 +157,14 @@ ContentContext::ContentContext(std::shared_ptr<Context> context)
       CreateDefaultPipeline<LinearGradientFillPipeline>(*context_);
   radial_gradient_fill_pipelines_[{}] =
       CreateDefaultPipeline<RadialGradientFillPipeline>(*context_);
+#ifndef FML_OS_ANDROID
```
Comment (jonahwilliams) [inline impeller/entity/contents/content_context.cc:160]: I don't really love this pattern. I think we may eventually need better scaffolding for backend specific shaders

**Comment by jonahwilliams** [comment]
I have mixed feelings about how I've currently handled the lack of support for this on Android, but also zoom zoom go fast.

**Comment by jonahwilliams** [comment]
Converting to draft, we should use SSBOs for this

File: impeller/entity/BUILD.gn [hunk 2]
```diff
@@ -66,6 +66,20 @@ impeller_shaders("entity_shaders") {
   ]
 }
 
+impeller_shaders("modern_entity_shaders") {
```
Comment (jonahwilliams) [inline impeller/entity/BUILD.gn:69]: This _almost_ works, except that I fail to load the fixed_fill shaders at runtime. I suspect something somewhere is hard coded but I haven't found it. If I place these shaders in the entity shaders set (and force the gles_language version to 460 everywhere) it works as expected

File: impeller/entity/contents/content_context.cc [hunk 3]
```diff
@@ -150,13 +150,26 @@ ContentContext::ContentContext(std::shared_ptr<Context> context)
   if (!context_ || !context_->IsValid()) {
     return;
   }
+#ifdef FML_OS_ANDROID
```
Comment (jonahwilliams) [inline impeller/entity/contents/content_context.cc:153]: TBD: actual feature selection when we have vulkan support

Comment (dnfield) [reply impeller/entity/contents/content_context.cc:153]: Consider having a define for `IMPELLER_SUPPORT_SSBO` that's just `IMPELLER_ENABLE_METAL || IMPELLER_ENABLE_VULKAN`. Or just use the more verbose one here. You'll need some runtime detection, so an update to the Context object to have some method or struct or something that says whether SSBOs are allowed.

Comment (dnfield) [reply impeller/entity/contents/content_context.cc:153]: Oh I see some of the runtime stuff is started.

Comment (jonahwilliams) [reply impeller/entity/contents/content_context.cc:153]: We need to leave this as runtime support as far as I can tell, once Vulkan is on by default. Avoids literring the code base with ifdefs too

Comment (chinmaygarde) [reply impeller/entity/contents/content_context.cc:153]: Can we have this on on `impeller::Context` instead? You can ask the context for its feature set. That way, only ES will return the legacy stuff.

File: ci/licenses_golden/licenses_flutter [hunk 4]
```diff
@@ -1285,6 +1286,7 @@ FILE: ../../../flutter/impeller/entity/shaders/glyph_atlas_sdf.frag
 FILE: ../../../flutter/impeller/entity/shaders/glyph_atlas_sdf.vert
 FILE: ../../../flutter/impeller/entity/shaders/gradient_fill.vert
 FILE: ../../../flutter/impeller/entity/shaders/linear_gradient_fill.frag
+FILE: ../../../flutter/impeller/entity/shaders/linear_gradient_fixed_fill.frag
```
Comment (chinmaygarde) [inline ci/licenses_golden/licenses_flutter:1289]: lets call this _ssbo_fill_ perhaps?

File: impeller/compiler/compiler.cc [hunk 5]
```diff
@@ -45,10 +45,14 @@ static CompilerBackend CreateGLSLCompiler(const spirv_cross::ParsedIR& ir,
   sl_options.force_zero_initialized_variables = true;
   sl_options.vertex.fixup_clipspace = true;
   if (source_options.target_platform == TargetPlatform::kOpenGLES) {
-    sl_options.version = 100;
+    sl_options.version = source_options.gles_language_version > 0
```
Comment (chinmaygarde) [inline impeller/compiler/compiler.cc:48]: I wonder if we should do source language selection based on #version directives now instead of applying command line options. Omitting them was fine when we only had a single version.

File: impeller/geometry/geometry_unittests.h [hunk 6]
```diff
@@ -141,3 +158,4 @@ inline ::testing::AssertionResult ColorBufferNear(
 #define ASSERT_SIZE_NEAR(a, b) ASSERT_PRED2(&::SizeNear, a, b)
 #define ASSERT_ARRAY_4_NEAR(a, b) ASSERT_PRED2(&::Array4Near, a, b)
 #define ASSERT_COLOR_BUFFER_NEAR(a, b) ASSERT_PRED2(&::ColorBufferNear, a, b)
+#define ASSERT_COLORS_NEAR(a, b) ASSERT_PRED2(&::ColorsNear, a, b)
```
Comment (chinmaygarde) [inline impeller/geometry/geometry_unittests.h:161]: Newline at EOF.

**Comment by chinmaygarde** [comment]
For clarity, where we support SSBOs, there is no case where we need to use the texture upload thingy right?

**Comment by jonahwilliams** [comment]
> For clarity, where we support SSBOs, there is no case where we need to use the texture upload thingy right?

Not that I can tell.

File: impeller/geometry/gradient.cc [hunk 7]
```diff
@@ -86,16 +89,90 @@ GradientData CreateGradientBuffer(const std::vector<Color>& colors,
...
+    // Use original buffer.
+    return std::nullopt;
+  }
+
+  uint32_t color_count;
+  auto minimum_delta = 1.0;
+  for (size_t i = 1; i < stops.size(); i++) {
+    auto value = stops[i] - stops[i - 1];
+    // Smaller than kEhCloseEnough
+    if (value < 0.0001) {
```
Comment (chinmaygarde) [inline impeller/geometry/gradient.cc:116]: Curious, why not just use the constant instead of hardcoding the value?

**Comment by chinmaygarde** [comment]
Using purely the SSBO approach on Metal and following up with the jumbo textures thing on legacy backends sounds good to me.

I reviewed the rest of the patch and it looks good to me barring the comments (mostly nits) already added. lgtm after it moves out of draft status.

File: impeller/geometry/gradient.cc (hunk 7 above)
Comment (jonahwilliams) [reply impeller/geometry/gradient.cc:116]: context was sort of https://github.com/flutter/engine/pull/36218

TLDR, if you have really close stops you need a smaller number

File: impeller/entity/contents/content_context.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/entity/contents/content_context.cc]: Done

File: impeller/entity/contents/backend_features.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  bool ssbo_support;
+};
+
+/// @brief feature sets available on most but not all modern hardware.
+constexpr BackendFeatures kModernBackendFeatures = {.ssbo_support = true};
+
+/// @brief Lowest common denominator feature sets.
+constexpr BackendFeatures kLegacyBackendFeatures = {.ssbo_support = false};
+
+}  // namespace impeller
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/entity/contents/backend_features.h]: Done

File: impeller/entity/contents/backend_features.h [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#pragma once
+
+namespace impeller {
+
+/// @brief A struct for describing available backend features for runtime
+/// selection.
+struct BackendFeatures {
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/entity/contents/backend_features.h]: Done

File: ci/licenses_golden/licenses_flutter (hunk 4 above)
Comment (jonahwilliams) [reply ci/licenses_golden/licenses_flutter:1289]: Done

File: impeller/compiler/compiler.cc (hunk 5 above)
Comment (jonahwilliams) [reply impeller/compiler/compiler.cc:48]: I think that sounds reasonable, but I'm not quite sure how to make make shaderc and spirvcross work together on this. I'll file a bug

File: impeller/geometry/geometry_unittests.h (hunk 6 above)
Comment (jonahwilliams) [reply impeller/geometry/geometry_unittests.h]: Done

**Comment by jonahwilliams** [comment]
Needs more tests still, and need to figure out why I can't find the shaders at runtime

File: impeller/geometry/gradient.cc (hunk 7 above)
Comment (jonahwilliams) [reply impeller/geometry/gradient.cc:116]: in the future, we'll need to stop computing the gradients on the CPU for certain stop values

File: impeller/compiler/compiler.cc (hunk 5 above)
Comment (jonahwilliams) [reply impeller/compiler/compiler.cc:48]: https://github.com/flutter/flutter/issues/115587

**Comment by jonahwilliams** [comment]
Fun times:

```
gen/flutter/impeller/entity/radial_gradient_ssbo_fill.frag.cc(22,1): error: static assertion failed due to requirement 'sizeof(impeller::RadialGradientSsboFillFragmentShader::ColorData<0>) == 0'
static_assert(sizeof(Shader::ColorData<0>) == 0);
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gen/flutter/impeller/entity/radial_gradient_ssbo_fill.frag.cc(22,44): note: expression evaluates to '4 == 0'
static_assert(sizeof(Shader::ColorData<0>) == 0);
```

File: impeller/compiler/code_gen_template.h [hunk 10]
```diff
@@ -202,7 +202,6 @@ using Shader = {{camel_case(shader_name)}}{{camel_case(shader_stage)}}Shader;
 // Sanity checks for {{def.name}}
 {% if last(def.members).array_elements == 0 %}
 static_assert(std::is_standard_layout_v<Shader::{{def.name}}<0>>);
-static_assert(sizeof(Shader::{{def.name}}<0>) == {{def.byte_length}});
```
Comment (jonahwilliams) [inline impeller/compiler/code_gen_template.h:205]: This assert appears to be invalid on windows hosts, which claim this always has a sizeof 4.

File: impeller/entity/BUILD.gn (hunk 2 above)
Comment (jonahwilliams) [reply impeller/entity/BUILD.gn:69]: Fixed

**Comment by jonahwilliams** [comment]
This is ready 😄

**Review by chinmaygarde — APPROVED** [review]

**Comment by jonahwilliams** [comment]
Yes all of the generated code worked correctly. Though TBH I didn't find a way to use the struct generated for the SSBO, since it is dynamically sized - so I ended up creating a BufferView myself. I think that is OK though
